pkg/kv/kvserver/mvcc_gc_queue.go

```diff
@@ -423,15 +458,35 @@ func makeMVCCGCQueueScoreImpl(
 		r.FinalScore++
 	}
 
-	// If range was deleted, lower score threshold to allow faster cleanup as
-	// most of the garbage is concentrated at the ttl threshold.
-	if !r.ShouldQueue && suspectedFullRangeDeletion(ms) {
+	maybeRangeDel := suspectedFullRangeDeletion(ms)
+	hasActiveGCHint := gcHintedRangeDelete(hint, gcTTL, now)
+
+	if hasActiveGCHint && (maybeRangeDel || ms.ContainsEstimates > 0) {
+		// We have GC hint allowing us to collect range and we either satisfy
+		// heuristic that indicate no live data or we have estimates and we assume
+		// hint is correct.
+		r.ShouldQueue = canAdvanceGCThreshold
+		r.FinalScore = deleteRangePriority
+	}
+
+	if !r.ShouldQueue && maybeRangeDel {
+		// If we don't have GC hint, but range del heuristic is satisfied, then
+		// check with lowered score threshold as all the data is deleted.
 		r.ShouldQueue = canAdvanceGCThreshold && r.FuzzFactor*valScore > mvccGCDropRangeKeyScoreThreshold
 	}
 
 	return r
 }
 
+func gcHintedRangeDelete(hint roachpb.GCHint, ttl time.Duration, now hlc.Timestamp) bool {
+	deleteTimestamp := hint.LatestRangeDeleteTimestamp
+	if deleteTimestamp.IsEmpty() {
+		return false
+	}
+	gcThreshold := deleteTimestamp.Add(ttl.Nanoseconds(), 0)
+	return gcThreshold.Less(now)
```

Comment:
nit: this isn't the GC threshold, so it's a bit confusing to refer to it as such. I don't think we even need a variable here, we can just do `deleteTimestamp.Add().Less(now)`.

Reply:
It would be better to say something like (now - ttl) > delete_time but code for it looks bad.